lib/Sema/TypeCheckRegex.cpp

```diff
@@ -0,0 +1,95 @@
+//===--- TypeCheckRegex.cpp - Regex type checking utilities ---------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "TypeCheckRegex.h"
+
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Type.h"
+#include "swift/AST/Types.h"
+
+using namespace swift;
+
+// Encoding rules:
+// 〚`T` (atom)〛 ==> .atom
+// 〚`name: T` (atom)〛 ==> .atom, `name`, '\0'
+// 〚`[T]`〛 ==> 〚`T`〛, .formOptional
+// 〚`T?`〛 ==> 〚`T`〛, .formOptional
+// 〚`(T0, T1, ...)`〛 ==> .beginTuple, 〚`T0`〛, 〚`T1`〛, ..., .endTuple
+//
+// For details, see apple/swift-experimental-string-processing.
+bool swift::decodeRegexCaptureTypes(ASTContext &ctx,
```

Comment:
This might be needed for short-term reasons, but is there any way we can have the package do the decoding by passing in function pointers for type constructors?

Reply:
That should be totally possible but would add complexity and make it hard to debug without a lot of mocking. Note we would also need to pass an ASTContext pointer to Swift and then back to the type constructors.